core/src/storage2/alloc/init.rs

```diff
@@ -0,0 +1,160 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::DynamicAllocator;
+use crate::storage2::traits::pull_spread_root;
+use cfg_if::cfg_if;
+use ink_primitives::Key;
+
+/// The default dynamic allocator key offset.
+///
+/// This is where the dynamic allocator is stored on the contract storage.
+const DYNAMIC_ALLOCATOR_KEY_OFFSET: [u8; 32] = [0xFE; 32];
+
+/// The phase in which a contract execution can be.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum ContractPhase {
+    /// Initializes the global dynamic storage allocator from scratch.
+    ///
+    /// Upon initialization it will be created from scratch as if the
+    /// contract has been deployed for the first time.
+    Deploy,
+    /// Initializes the global dynamic storage allocator from storage.
+    ///
+    /// Upon initialization the dynamic storage allocator will be pulled
+    /// from the contract storage with the assumption that a former
+    /// contract deployment has already taken place in the past.
+    Call,
+}
+
+/// The state of the dynamic allocator global instance.
+#[derive(Debug)]
+#[allow(clippy::large_enum_variant)]
+enum DynamicAllocatorState {
+    /// The global instance has not yet been initialized.
+    ///
+    /// Upon initialization it will be created from scratch as if the
+    /// contract has been deployed for the first time.
+    UninitDeploy,
+    /// The global instance has not yet been initialized.
+    ///
+    /// Upon initialization it will be pulled from the contract storage
+    /// with the assumption that a former contract deployment has already
+    /// taken place in the past.
+    UninitCall,
+    /// The global instance has already been initialized successfully.
+    Initialized(DynamicAllocator),
+}
+
+impl From<ContractPhase> for DynamicAllocatorState {
+    fn from(phase: ContractPhase) -> Self {
+        match phase {
+            ContractPhase::Deploy => DynamicAllocatorState::UninitDeploy,
+            ContractPhase::Call => DynamicAllocatorState::UninitCall,
+        }
+    }
+}
+
+cfg_if! {
+    if #[cfg(all(not(feature = "std"), target_arch = "wasm32"))] {
+        // Procedures for the Wasm compilation:
+
+        /// The global instance for the dynamic storage allocator.
+        static mut GLOBAL_INSTANCE: DynamicAllocatorState = DynamicAllocatorState::UninitDeploy;
+
+        /// Commands the (re-)initialization of the global instance for the dynamic
+        /// storage allocator.
+        pub fn initialize_for(phase: ContractPhase) {
+            let instance = unsafe { &mut GLOBAL_INSTANCE };
+            // We do not allow reinitialization for Wasm targets for performance reasons.
+            if let DynamicAllocatorState::Initialized(_) = instance {
+                panic!("cannot reinitialize dynamic storage allocator instance in Wasm");
+            }
+            *instance = phase.into();
```

Comment:
Under what circumstances would we go from `UninitCall` back to `UninitDeploy`?

Reply:
Never in on-chain mode. In off-chain mode we can do this if we execute tests in a single thread which happens on for example `miri`.